hibernate-core/src/test/java/org/hibernate/orm/test/bytecode/enhancement/lazy/LazyLoadingAndParameterizedInheritanceTest.java

```diff
@@ -0,0 +1,191 @@
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.orm.test.bytecode.enhancement.lazy;
+
+import jakarta.persistence.Entity;
+import jakarta.persistence.FetchType;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.Id;
+import jakarta.persistence.JoinColumn;
+import jakarta.persistence.ManyToOne;
+import jakarta.persistence.MappedSuperclass;
+import jakarta.persistence.OneToMany;
+
+import org.hibernate.cfg.AvailableSettings;
+
+import org.hibernate.testing.bytecode.enhancement.extension.BytecodeEnhanced;
+import org.hibernate.testing.orm.junit.DomainModel;
+import org.hibernate.testing.orm.junit.JiraKey;
+import org.hibernate.testing.orm.junit.ServiceRegistry;
+import org.hibernate.testing.orm.junit.SessionFactory;
+import org.hibernate.testing.orm.junit.SessionFactoryScope;
+import org.hibernate.testing.orm.junit.Setting;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@DomainModel(
+		annotatedClasses = {
+				LazyLoadingAndParameterizedInheritanceTest.One.class,
+				LazyLoadingAndParameterizedInheritanceTest.AbsOne.class,
+				LazyLoadingAndParameterizedInheritanceTest.AbsTwo.class,
+				LazyLoadingAndParameterizedInheritanceTest.Three.class,
+				LazyLoadingAndParameterizedInheritanceTest.Two.class,
+		}
+)
+@ServiceRegistry(
+		settings = {
+				@Setting(name = AvailableSettings.USE_SECOND_LEVEL_CACHE, value = "false"),
+				@Setting(name = AvailableSettings.ENABLE_LAZY_LOAD_NO_TRANS, value = "true"),
+		}
+)
+@SessionFactory
+@BytecodeEnhanced
+@JiraKey("HHH-18151")
+public class LazyLoadingAndParameterizedInheritanceTest {
+
+	private Long oneId;
+	private Long threeId;
+
+	@BeforeEach
+	public void prepare(SessionFactoryScope scope) {
+		scope.inTransaction( s -> {
+			final var one = new One();
+			final var two = new Two();
+			final var three = new Three();
+			one.setTwo( two );
+			two.getOnes().add( one );
+			two.setThree( three );
+			three.getTwos().add( two );
+
+			s.persist( one );
+			s.persist( two );
+			s.persist( three );
+			oneId = one.getId();
+			threeId = three.getId();
+		} );
+	}
+
+	@Test
+	public void test(SessionFactoryScope scope) {
+		scope.inTransaction( s -> {
+			One one = s.find( One.class, oneId );
+			Two two = one.getTwo();
+			final var three = two.getThree();
+
+			final var actualThree = s.find( Three.class, threeId );
+			assertThat( actualThree ).isNotNull();
+			assertThat( actualThree.getTwos().iterator().next().getId() ).isEqualTo( two.getId() );
+
+			//That is the actual test. If three == null ==> lazy load was not performed.
+			assertThat( three ).isNotNull();
+		} );
+	}
+
+	@Entity
+	public static class One extends AbsOne<Two> {
```

Comment:
It's generally a good idea to use explicit entity names for inner classes to avoid long table names (unsupported by some old db versions we only test on the nightly CI). Please also do that for the other entities in the test.

```suggestion
	@Entity( name = "One" )
	public static class One extends AbsOne<Two> {
```